File: electron/agent/tools/glob.ts
/**
 * glob tool — find files by pattern.
 *
 * Dedicated tool (not a shell `find` wrapper) so the model gets structured
 * results instead of freeform command output. Supports the common glob
 * syntax: *, **, ?, character classes. Respects a basic .gitignore
 * (node_modules, .git, dist) by default.
 */

import * as fs from 'fs';
import * as path from 'path';
import { tool } from 'ai';
import { z } from 'zod';
import { resolveAndFollowSymlinks } from '../path-safety';
import type { ToolResult, ToolRegistration } from './types';
import type { ToolContext } from './tool-context';
import { withPermission } from '../permission-wrapper';

const MAX_RESULTS = 200;
const IGNORE_DIRS = new Set(['node_modules', '.git', 'dist', 'dist-electron', 'release', '.next', '.cache']);

/** Shared body — takes the parsed args + workspaceRoot (the only ctx field it needs). */
export async function runGlob(
  pattern: string,
  relPath: string,
  workspaceRoot: string,
): Promise<ToolResult> {
  if (!pattern) return { status: 'failed', output: 'Missing required arg: pattern' };

  let root: string;
  try {
    root = relPath
      ? resolveAndFollowSymlinks(workspaceRoot, relPath)
      : workspaceRoot;
  } catch (e: any) {
    return { status: 'failed', output: `Path error: ${e.message}` };
  }

  try {
    const stats = fs.statSync(root);
    if (!stats.isDirectory()) {
      return { status: 'failed', output: `Not a directory: ${relPath || '(root)'}` };
    }
  } catch {
    return { status: 'failed', output: `Directory not found: ${relPath || '(root)'}` };
  }

  const matches: string[] = [];
  const regex = globToRegex(pattern);
  walk(root, '', (rel) => {
    if (matches.length >= MAX_RESULTS) return false;
    // Normalize to forward slashes for matching + display.
    const normalized = rel.split(path.sep).join('/');
    if (regex.test(normalized)) {
      matches.push(normalized);
    }
    return true;
  });

  if (matches.length === 0) {
    return {
      status: 'executed',
      output: `No files matching "${pattern}" in ${relPath || '.'}.`,
      meta: '0 matches',
      display: { kind: 'file_list', paths: [] },
    };
  }

  matches.sort();
  return {
    status: 'executed',
    output: `${matches.length} match${matches.length === 1 ? '' : 'es'} for "${pattern}":\n${matches.slice(0, 50).join('\n')}${matches.length > 50 ? `\n…and ${matches.length - 50} more` : ''}`,
    meta: `${matches.length} files`,
    display: { kind: 'file_list', paths: matches },
  };
}

export const globTool: ToolRegistration = {
  name: 'glob',
  definition: {
    name: 'glob',
    description:
      'Find files matching a glob pattern. Supports * (single segment), ** (any depth), ' +
      '? (single char), and [abc] (char class). Returns up to 200 paths relative to ' +
      'the workspace root. Ignores node_modules/.git/dist by default. Faster than ' +
      'list_dir when you know the extension or naming pattern.',
    input_schema: {
      type: 'object',
      properties: {
        pattern: {
          type: 'string',
          description: 'Glob pattern, e.g. "src/**/*.tsx", "**/*.test.ts", "lib/*.md".',
        },
        path: {
          type: 'string',
          description: 'Subdirectory to search in (relative to workspace root). Defaults to workspace root.',
        },
      },
      required: ['pattern'],
    },
  },
  riskTier: 'read_only',
  requiresWorktree: false,
  timeoutMs: 10_000,
  autoApproveIn: ['plan', 'ask', 'edit', 'full'],
  execute: async (args, ctx) =>
    runGlob(String(args.pattern ?? ''), typeof args.path === 'string' ? args.path : '', ctx.workspaceRoot),
};

// ─── SDK factory (Phase 2) ─────────────────────────────────────────────

export function createGlobTool(ctx: ToolContext) {
  return tool({
    description:
      'Find files matching a glob pattern. Supports * (single segment), ** (any depth), ' +
      '? (single char), and [abc] (char class). Returns up to 200 paths relative to ' +
      'the workspace root. Ignores node_modules/.git/dist by default. Faster than ' +
      'list_dir when you know the extension or naming pattern.',
    inputSchema: z.object({
      pattern: z.string().describe('Glob pattern, e.g. "src/**/*.tsx", "**/*.test.ts", "lib/*.md".'),
      path: z.string().optional().describe('Subdirectory to search in (relative to workspace root). Defaults to workspace root.'),
    }),
    execute: async ({ pattern, path }) =>
      withPermission(ctx, 'glob', { pattern, path }, () => runGlob(pattern, path ?? '', ctx.workspaceRoot)),
  });
}

/** Walk a directory tree, calling visitor(relPath) for every file. Visitor
 *  returns false to stop the walk (used for the MAX_RESULTS cap). */
function walk(rootAbs: string, relDir: string, visit: (relPath: string) => boolean): void {
  let entries: fs.Dirent[];
  try {
    entries = fs.readdirSync(rootAbs, { withFileTypes: true });
  } catch {
    return;
  }
  for (const entry of entries) {
    if (entry.isDirectory()) {
      if (IGNORE_DIRS.has(entry.name)) continue;
      const childRel = relDir ? `${relDir}/${entry.name}` : entry.name;
      const childAbs = path.join(rootAbs, entry.name);
      walk(childAbs, childRel, visit);
    } else if (entry.isFile()) {
      const rel = relDir ? `${relDir}/${entry.name}` : entry.name;
      if (!visit(rel)) return;
    }
  }
}

/** Convert a glob pattern to a RegExp. Supports *, **, ?, [abc]. */
function globToRegex(pattern: string): RegExp {
  let re = '';
  let i = 0;
  while (i < pattern.length) {
    const c = pattern[i];
    if (c === '*' && pattern[i + 1] === '*') {
      // ** — match any number of path segments (including zero).
      i += 2;
      if (pattern[i] === '/') i++;
      re += '.*';
    } else if (c === '*') {
      // * — match within a single path segment (not /).
      re += '[^/]*';
      i++;
    } else if (c === '?') {
      re += '[^/]';
      i++;
    } else if (c === '[') {
      // Pass through character classes.
      const end = pattern.indexOf(']', i);
      if (end === -1) { re += '\\['; i++; }
      else { re += pattern.slice(i, end + 1); i = end + 1; }
    } else if ('.+^${}()|\\'.includes(c)) {
      re += '\\' + c;
      i++;
    } else {
      re += c;
      i++;
    }
  }
  return new RegExp(`^${re}$`);
}

File: src-tauri/crates/tide-tools/src/tools/glob.rs
//! glob — port of `app/core/agent/tools/glob.ts` (). Find files by
//! pattern (`*`, `**`, `?`, `[abc]`), capped at 200 results, ignoring the
//! common build/dependency dirs. Returns workspace-relative, forward-slash
//! paths.

use std::path::Path;

use regex::Regex;
use serde_json::json;

use crate::path_safety::resolve_and_follow_symlinks;
use crate::permission::RiskTier;
use crate::{Tool, ToolContext, ToolDisplay, ToolError, ToolOutcome, ToolSpec};

use super::arg_str;

pub(crate) const MAX_RESULTS: usize = 200;
const IGNORE_DIRS: &[&str] = &[
    "node_modules",
    ".git",
    "dist",
    "dist-electron",
    "release",
    ".next",
    ".cache",
];

const DESCRIPTION: &str = "Find files matching a glob pattern. Supports * (single segment), ** (any depth), ? (single char), and [abc] (char class). Returns up to 200 paths relative to the workspace root. Ignores node_modules/.git/dist by default. Faster than list_dir when you know the extension or naming pattern.";

pub struct GlobTool;

pub(crate) fn run_glob(pattern: &str, rel_path: &str, workspace_root: &Path) -> ToolOutcome {
    if pattern.is_empty() {
        return ToolOutcome::failed("Missing required arg: pattern");
    }

    let root = if rel_path.is_empty() {
        workspace_root.to_path_buf()
    } else {
        match resolve_and_follow_symlinks(workspace_root, rel_path) {
            Ok(root) => root,
            Err(e) => return ToolOutcome::failed(format!("Path error: {e}")),
        }
    };

    let meta = match std::fs::metadata(&root) {
        Ok(m) => m,
        Err(_) => {
            return ToolOutcome::failed(format!(
                "Directory not found: {}",
                if rel_path.is_empty() { "(root)" } else { rel_path }
            ))
        }
    };
    if !meta.is_dir() {
        return ToolOutcome::failed(format!(
            "Not a directory: {}",
            if rel_path.is_empty() { "(root)" } else { rel_path }
        ));
    }

    let regex = glob_to_anchored_regex(pattern);
    let mut matches: Vec<String> = Vec::new();
    walk(&root, "", &mut |rel| {
        if matches.len() >= MAX_RESULTS {
            return false;
        }
        // Normalize to forward slashes for matching + display.
        let normalized = rel.replace(std::path::MAIN_SEPARATOR, "/");
        if regex.is_match(&normalized) {
            matches.push(normalized);
        }
        true
    });

    if matches.is_empty() {
        return ToolOutcome::executed(format!(
            "No files matching \"{pattern}\" in {}.",
            if rel_path.is_empty() { "." } else { rel_path }
        ))
        .with_meta("0 matches")
        .with_display(ToolDisplay::FileList { paths: Vec::new() });
    }

    matches.sort();
    let shown: Vec<&str> = matches.iter().take(50).map(|s| s.as_str()).collect();
    let mut output = format!(
        "{} match{} for \"{pattern}\":\n{}",
        matches.len(),
        if matches.len() == 1 { "" } else { "es" },
        shown.join("\n")
    );
    if matches.len() > 50 {
        output.push_str(&format!("\n…and {} more", matches.len() - 50));
    }
    ToolOutcome::executed(output)
        .with_meta(format!("{} files", matches.len()))
        .with_display(ToolDisplay::FileList {
            paths: matches.clone(),
        })
}

/// Walk collecting file paths relative to `root` (forward slashes), with
/// the TS visitor-returns-false early stop. Dotfiles ARE visited; only the
/// IGNORE_DIRS set is skipped (glob.ts semantics — unlike grep's walker).
fn walk(dir: &Path, rel_dir: &str, visit: &mut dyn FnMut(&str) -> bool) {
    let entries = match std::fs::read_dir(dir) {
        Ok(e) => e,
        Err(_) => return,
    };
    for entry in entries.flatten() {
        let name = entry.file_name().to_string_lossy().to_string();
        let Ok(file_type) = entry.file_type() else { continue };
        if file_type.is_dir() {
            if IGNORE_DIRS.contains(&name.as_str()) {
                continue;
            }
            let child_rel = if rel_dir.is_empty() {
                name.clone()
            } else {
                format!("{rel_dir}/{name}")
            };
            walk(&entry.path(), &child_rel, visit);
        } else if file_type.is_file() {
            let rel = if rel_dir.is_empty() {
                name.clone()
            } else {
                format!("{rel_dir}/{name}")
            };
            if !visit(&rel) {
                return;
            }
        }
    }
}

/// Glob pattern → anchored regex, ported from glob.ts `globToRegex`:
/// `**` (consuming one following `/`) spans any depth, `*` stays within a
/// segment, `?` is one char, `[abc]` passes through, specials are escaped.
fn glob_to_anchored_regex(pattern: &str) -> Regex {
    let mut re = String::new();
    let mut chars = pattern.chars().peekable();
    while let Some(c) = chars.next() {
        match c {
            '*' => {
                if chars.peek() == Some(&'*') {
                    chars.next();
                    if chars.peek() == Some(&'/') {
                        chars.next();
                    }
                    re.push_str(".*");
                } else {
                    re.push_str("[^/]*");
                }
            }
            '?' => re.push_str("[^/]"),
            '[' => {
                // Pass character classes through; an unterminated '[' is literal.
                let rest: String = chars.clone().collect();
                if let Some(end) = rest.find(']') {
                    re.push('[');
                    re.push_str(&rest[..=end]);
                    for _ in 0..=end {
                        chars.next();
                    }
                } else {
                    re.push_str("\\[");
                }
            }
            c if ".+^${}()|\\".contains(c) => {
                re.push('\\');
                re.push(c);
            }
            c => re.push(c),
        }
    }
    let anchored = format!("^{re}$");
    Regex::new(&anchored).unwrap_or_else(|_| Regex::new("^$").unwrap())
}

impl Tool for GlobTool {
    fn spec(&self) -> ToolSpec {
        ToolSpec {
            name: "glob".into(),
            description: DESCRIPTION.into(),
            parameters: json!({
                "type": "object",
                "properties": {
                    "pattern": { "type": "string", "description": "Glob pattern, e.g. \"src/**/*.tsx\", \"**/*.test.ts\", \"lib/*.md\"." },
                    "path": { "type": "string", "description": "Subdirectory to search in (relative to workspace root). Defaults to workspace root." }
                },
                "required": ["pattern"]
            }),
        }
    }

    fn risk_tier(&self) -> RiskTier {
        RiskTier::ReadOnly
    }

    fn execute(&self, ctx: &ToolContext, args: serde_json::Value) -> Result<ToolOutcome, ToolError> {
        let pattern = arg_str(&args, "pattern");
        let path = arg_str(&args, "path");
        Ok(run_glob(&pattern, &path, &ctx.workspace_root))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn workspace() -> tempfile::TempDir {
        let tmp = tempfile::tempdir().unwrap();
        std::fs::create_dir_all(tmp.path().join("src/components/ui")).unwrap();
        std::fs::create_dir_all(tmp.path().join("node_modules/pkg")).unwrap();
        std::fs::create_dir_all(tmp.path().join("dist")).unwrap();
        std::fs::write(tmp.path().join("src/a.ts"), "x").unwrap();
        std::fs::write(tmp.path().join("src/components/b.tsx"), "x").unwrap();
        std::fs::write(tmp.path().join("src/components/ui/c.tsx"), "x").unwrap();
        std::fs::write(tmp.path().join("README.md"), "x").unwrap();
        std::fs::write(tmp.path().join("node_modules/pkg/d.js"), "x").unwrap();
        std::fs::write(tmp.path().join("dist/e.js"), "x").unwrap();
        tmp
    }

    #[test]
    fn missing_pattern_fails() {
        let tmp = workspace();
        let out = run_glob("", "", tmp.path());
        assert_eq!(out.status, crate::OutcomeStatus::Failed);
        assert!(out.output.contains("Missing required arg"));
    }

    #[test]
    fn star_within_segment_only() {
        let tmp = workspace();
        let out = run_glob("src/*.ts", "", tmp.path());
        let ToolDisplay::FileList { paths } = out.display.unwrap() else {
            panic!("file_list display");
        };
        assert_eq!(paths, vec!["src/a.ts"]);
    }

    #[test]
    fn double_star_spans_depths() {
        let tmp = workspace();
        let out = run_glob("src/**/*.tsx", "", tmp.path());
        let ToolDisplay::FileList { paths } = out.display.unwrap() else {
            panic!("file_list display");
        };
        assert_eq!(paths, vec!["src/components/b.tsx", "src/components/ui/c.tsx"]);
        assert_eq!(out.meta.as_deref(), Some("2 files"));
        assert!(out.output.contains("2 matches for \"src/**/*.tsx\""));
    }

    #[test]
    fn question_mark_and_char_class() {
        let tmp = workspace();
        let out = run_glob("?.?", "", tmp.path());
        let ToolDisplay::FileList { paths } = out.display.unwrap() else {
            panic!("file_list display");
        };
        assert!(paths.is_empty());

        let out = run_glob("READ[M]E.md", "", tmp.path());
        let ToolDisplay::FileList { paths } = out.display.unwrap() else {
            panic!("file_list display");
        };
        assert_eq!(paths, vec!["README.md"]);
    }

    #[test]
    fn ignores_common_dirs() {
        let tmp = workspace();
        let out = run_glob("**/*.js", "", tmp.path());
        let ToolDisplay::FileList { paths } = out.display.unwrap() else {
            panic!("file_list display");
        };
        assert!(paths.is_empty(), "node_modules/dist must be ignored: {paths:?}");
        assert!(out.output.starts_with("No files matching"));
        assert_eq!(out.meta.as_deref(), Some("0 matches"));
    }

    #[test]
    fn results_sorted_and_capped_display() {
        let tmp = tempfile::tempdir().unwrap();
        for i in 0..60 {
            std::fs::write(tmp.path().join(format!("f{i:02}.txt")), "x").unwrap();
        }
        let out = run_glob("*.txt", "", tmp.path());
        assert!(out.output.contains("60 matches"));
        assert!(out.output.contains("…and 10 more"));
        let ToolDisplay::FileList { paths } = out.display.unwrap() else {
            panic!("file_list display");
        };
        assert_eq!(paths.len(), 60);
        let mut sorted = paths.clone();
        sorted.sort();
        assert_eq!(paths, sorted);
    }

    #[test]
    fn path_scope_resolves_subdir() {
        let tmp = workspace();
        let out = run_glob("**/*.tsx", "src/components", tmp.path());
        let ToolDisplay::FileList { paths } = out.display.unwrap() else {
            panic!("file_list display");
        };
        assert_eq!(paths, vec!["b.tsx", "ui/c.tsx"]);
    }

    #[test]
    fn missing_dir_and_traversal_fail() {
        let tmp = workspace();
        let out = run_glob("*", "ghost", tmp.path());
        assert_eq!(out.status, crate::OutcomeStatus::Failed);
        assert!(out.output.contains("Directory not found"));
        let out = run_glob("*", "../../etc", tmp.path());
        assert!(out.output.contains("Path error") || out.output.contains("Directory not found"));
    }

    #[test]
    fn execute_routes_through_trait() {
        let tmp = workspace();
        let tool = GlobTool;
        assert_eq!(tool.spec().name, "glob");
        assert_eq!(tool.risk_tier(), RiskTier::ReadOnly);
        let out = tool
            .execute(
                &ToolContext::new(tmp.path().to_path_buf()),
                json!({ "pattern": "*.md" }),
            )
            .unwrap();
        assert!(out.output.contains("README.md"));
    }
}
